### ISSUE TITLE
[naming-convention] Can't use array of selectors

### ISSUE THREAD

**Issue by EvgenyOrekhov** [typescript-eslint/typescript-eslint#2376]
eslint-plugin 3.8.0 introduced the ability to specify `selector` as an array. But it doesn't actually work.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": ["variable", "parameter", "property", "accessor"],
        "types": ["boolean"],
        "format": ["PascalCase"],
        "prefix": ["is", "should", "has", "can", "did", "will"]
      }
    ]
  }
}
```

```TS
// any code
```

**Expected Result**

Rule should not throw.

**Actual Result**

Rule throws 

```text
Configuration for rule "@typescript-eslint/naming-convention" is invalid:
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value ["variable","parameter","property","accessor"] should be string.
        Value ["variable","parameter","property","accessor"] should be equal to one of the allowed values.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value ["variable","parameter","property","accessor"] should be string.
        Value ["variable","parameter","property","accessor"] should be equal to one of the allowed values.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value ["variable","parameter","property","accessor"] should be string.
        Value ["variable","parameter","property","accessor"] should be equal to one of the allowed values.
        Value ["variable","parameter","property","accessor"] should be string.
        Value ["variable","parameter","property","accessor"] should be equal to one of the allowed values.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value ["variable","parameter","property","accessor"] should be string.
        Value ["variable","parameter","property","accessor"] should be equal to one of the allowed values.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
        Value {"selector":["variable","parameter","property","accessor"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should match exactly one schema in oneOf.
```

**Additional Info**

Previously, I was using separate configuration for each selector, and it worked:

```json
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "variable",
        "types": ["boolean"],
        "format": ["PascalCase"],
        "prefix": ["is", "should", "has", "can", "did", "will"]
      },
      {
        "selector": "parameter",
        "types": ["boolean"],
        "format": ["PascalCase"],
        "prefix": ["is", "should", "has", "can", "did", "will"]
      },
      {
        "selector": "property",
        "types": ["boolean"],
        "format": ["PascalCase"],
        "prefix": ["is", "should", "has", "can", "did", "will"]
      },
      {
        "selector": "accessor",
        "types": ["boolean"],
        "format": ["PascalCase"],
        "prefix": ["is", "should", "has", "can", "did", "will"]
      }
    ]
  }
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.8.0` |
| `@typescript-eslint/parser`        | `3.8.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `14.6.0` |
| `npm`                              | `6.14.6` |

**Comment by bradzacher** [comment]
I'm unable to repro this against master.
There are passing tests with the array selector: https://github.com/typescript-eslint/typescript-eslint/blob/470174ad51fdb12d82129a896559075513f6c912/packages/eslint-plugin/tests/rules/naming-convention.test.ts#L730-L750

I also tried against 3.8.0 and it worked fine.

Finally I tried your exact config against the raw config schema and it worked fine: https://www.jsonschemavalidator.net/s/S63J8Yol

Likely you do not have the correct version installed, or have multiple versions installed.

**Comment by EvgenyOrekhov** [comment]
@bradleyayers I have a single version installed, and it is 3.8.0.

Here is a reproducible case (not a minimal case though, it's a real case that I want in my project, but the original minimal case is also reproducible with my setup): https://github.com/EvgenyOrekhov/eslint-config-hardcore/pull/124/files.

Here you can see that the version in `package-lock.json` is 3.8.0: https://github.com/EvgenyOrekhov/eslint-config-hardcore/blob/84e86e4e10820a7c653e3352e658e51aae0f2bd4/package-lock.json#L133.

And here is the error: https://github.com/EvgenyOrekhov/eslint-config-hardcore/pull/124/checks?check_run_id=964399541#step:5:14.

All dependencies, including transitive ones, are upgraded to latest versions.

**Comment by bradzacher** [comment]
I just cloned your repo and everything is working fine.
Seems like a problem on your end and with your CI.

Steps:
1) clone repo
1) patch your PR#124
1) npm i
1) run test: `npm test`
    - works fine
    ```shell
    bradzacher-mbp:eslint-config-hardcore-master bradzacher$ npm test
    
    > eslint-config-hardcore@12.6.0 test /Users/bradzacher/temp/eslint-config-hardcore-master
    > node test.js && eslint .
    ```
1) verify eslint does resolve config correctly:
`npx eslint --print-config test.js | jq '.rules["@typescript-eslint/naming-convention"]'`
    - works fine:

```shell
bradzacher-mbp:eslint-config-hardcore-master bradzacher$ npm test

> eslint-config-hardcore@12.6.0 test /Users/bradzacher/temp/eslint-config-hardcore-master
> node test.js && eslint .

bradzacher-mbp:eslint-config-hardcore-master bradzacher$ npx eslint --print-config test.js | jq '.rules["@typescript-eslint/naming-convention"]'
[
  "error",
  {
    "selector": "default",
    "format": [
      "strictCamelCase"
    ]
  },
  {
    "selector": "variable",
    "format": [
      "strictCamelCase",
      "UPPER_CASE"
    ]
  },
  {
    "selector": "typeLike",
    "format": [
      "StrictPascalCase"
    ]
  },
  {
    "selector": [
      "variable",
      "parameter",
      "property",
      "accessor"
    ],
    "format": [
      "PascalCase"
    ],
    "prefix": [
      "is",
      "should",
      "has",
      "can",
      "did",
      "will"
    ]
  }
]
```

**Comment by EvgenyOrekhov** [comment]
@bradzacher That is really weird. Could you tell me which node and npm versions did you use?

**Comment by bradzacher** [comment]
```shell
$ node -v
v12.14.1

$ npm -v
6.13.4
```

**Comment by EvgenyOrekhov** [comment]
I tried using Node 12.14.1 in my CI, it failed with the same error. I also tried installing `@typescript-eslint/parser` and `@typescript-eslint/eslint-plugin` v3.8.1-alpha.0, it also failed with the same error.

Also, I tried it on a different machine that uses Windows 10 instead of Ubuntu, and it also failed with the same error.

So, it is reproducible on my end on 3 different machines (Ubuntu 18.04.4 LTS, Ubuntu 20.04.1 LTS, Windows 10). Any suggestions?

**Comment by bradzacher** [comment]
The new version 100% supports the array of selectors.

The only way that you can get that error is if you have an older version of the package installed.
You'll have to figure out where / why it's installed on your machines, and fix that.

**Comment by bradzacher** [comment]
Okay, so the problem is that you can't currently use `types` and `selector`.
I manually patched your changes wrong, hence it worked fine.

**Comment by mdziekon** [comment]
Hi @bradzacher, you've just closed a duplicate of this issue (#2377), which is totally fine. However, I've included some analysis of the problem, a possible solution and a question about the proposed solution (or rather, the problem itself) that I think needs to be answered or at least discussed. Can you take a look at `Additional info` section over there and see if you can comment on that? I can copy-paste it here, if you want to maintain continuity.

**Comment by bradzacher** [comment]
The approach that has been taken for `selector` is to just not do schema validation anything because it's not possible to handle it.

Well, technically it is possible, but building the schema to support that would be a literal nightmare, so it's not worth it.

For modifiers - it doesn't matter, because the selectors only check for modifiers  they care about, so if you specify the wrong ones it just doesn't do anything.

Types OTOH are slightly different, because the `isCorrectType` function right now just assumes the config is correct.
So this function will need to be updated in some way to understand what selectors can have types checked.

The original PR that added support went a lazy route.
In the config normalisation step, the author just did a binary OR on the selector array to merge the many selectors into one. This worked fine at the time.

However this has the downside of making it harder to determine what selector is currently being validated. So as part of this fix, it will probably be better to revert that, and instead make the normalisation step produce 1 config per selector in the selector array.

That way you can easily determine what selector is currently being validated, and thus can easily tell if you should or should not apply the `types` for the selector.

**Comment by ginger-kang** [comment]
May I try this issue?

**Comment by bradzacher** [comment]
Feel free! Happy to accept a PR
